implementation/src/main/java/io/smallrye/config/SecretKeys.java

```diff
@@ -0,0 +1,51 @@
+package io.smallrye.config;
+
+import java.io.Serializable;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class SecretKeys implements Serializable {
+    private static final ThreadLocal<Boolean> LOCKED = ThreadLocal.withInitial(() -> Boolean.TRUE);
+
+    private final Function<String, Boolean> secrets;
+
+    public SecretKeys() {
+        this.secrets = (Function<String, Boolean> & Serializable) s -> false;
+    }
+
+    public SecretKeys(final Function<String, Boolean> secrets) {
+        this.secrets = secrets != null ? (Function<String, Boolean> & Serializable) secrets
+                : (Function<String, Boolean> & Serializable) s -> false;
+    }
+
+    public SecretKeys(final Set<String> secrets) {
+        this.secrets = (Function<String, Boolean> & Serializable) secrets::contains;
+    }
+
+    public boolean isSecret(String propertyName) {
+        return secrets.apply(propertyName);
+    }
+
+    public boolean isSecretAccessible(String propertyName) {
+        return !isSecret(propertyName) || !LOCKED.get();
+    }
+
+    public void accessSecrets(Runnable runnable) {
```

Comment:
These four methods should be `static`, and not dependent on the property name.  Rather I think it would make sense to have a query (`isLocked()`) that the interceptor checks.  If the name is in the set, and `isLocked()` is `true`, then it will simply throw `SecurityException`.